NEAR/MSI msi2isis and MsiCamera model improvements

KrisBecker opened the pull request:
NEAR/MSI support in ISIS has been improved and enhanced

## Description
The NEAR/MSI Eros data has been improved via a NASA PDART proposal. This work includes improvements in remediation of hydrazine contamination and in support of new global color maps.

Modifications have been made to the `msi2isis` application that will ingest the new dataset that will be archived in the PDS. The new images have already been expanded to the standard size image 537x412 where lines are interpolated from the original dataset size of 537x244. Images of expanded size are accepted as long as all the expected values in the labels are valid. In addition, a new option called TRIM has been added that allows the user choose to trim the edges of the images. The remediated dataset has shown these data regions to be improved sufficiently to retain for mapping purposes. And the SCLK times are modified to be valid with used with the NEAR/MSI SCLK kernel. This modification, along with corresponding changes in the MsiCamera model, provides support for the `sumspice` application to apply Gaskell SUMFILE ephemeris data updates.

The `MsiCamera` model has been modified to use the SpacecraftClockStartCount SCLK values rather than StartTime which is in UTC. This change is required for `sumspice` compatibly. Also, additional NAIF keywords were added to the ISIS IAK that provides support for improved geometry. These values are:

```
INS-93001_SWAP_OBSERVER_TARGET = 'TRUE'
INS-93001_LIGHTTIME_CORRECTION = 'LT+S'
INS-93001_LT_SURFACE_CORRECT   = 'FALSE'
```

This results in a new version of the NEAR/MSI IAK. Both of these IAKs are now stored/track in the repository in the $ISISROOT/src/near/objs/MsiCamera/iak directory. The new version, `msiAddendum002.ti`, needs to be copied to $ISISDATA/near/kernels/iak directory with this PR and prior to testing/release.

Note the modifications to the` MsiCamera` model require a new version number bump since new keywords are being read from image labels (`SpacecraftClockStartCount`) and IAK (as shown above).

The serial number configuration file, `NearMsiSerialNumber.trn`, has also been updated to accomodate inclusion of an optional keyword we use for control-to-ground strategies.

All `msi2isis` and `MsiCamera` tests have been converted to the ISIS Google Test framework. As such, the `msi2isis` application has been converted to a callable function. All old app and unit tests have been removed. This means that all test data files must be removed as well. Of note is I had to explicitly add (unignore) the NEAR/MIS *.lbl files to .gitignore as lbl files are ignored repo-wide.

NOTE that I did not update the application test as I am unsure how to get these changes into the system. I think it only requires an update to the truth files once the tests have been rerun. Please advise/assist.


## Related Issue
This PR address #4886.

## Motivation and Context
We have produced a new set of NEAR/MSI Eros images where the remediation of hydrazine contamination has been [improved](https://www.hou.usra.edu/meetings/lpsc2021/pdf/1682.pdf). To support this dataset, modifications and updates to the NEAR/MSI ingestion application, `msi2isis`, and the camera model were needed. These modifications are included in this PR.

## How Has This Been Tested?
The following shows results of the ctest suite for these changes:
```
(Mar282022) zion2[1374]: ctest -R Msi2isis                  
Test project /Users/kbecker/ISIS/Isis4Dev/Eros/MSI/Mar282022/ISIS3/build
      Start 1934: Msi2isis.Msi2isisTestGblIngestDefault
 1/13 Test #1934: Msi2isis.Msi2isisTestGblIngestDefault ..........   Passed    0.32 sec
      Start 1935: Msi2isis.Msi2isisTestGblIngestNoTrim
 2/13 Test #1935: Msi2isis.Msi2isisTestGblIngestNoTrim ...........   Passed    0.32 sec
      Start 1936: Msi2isis.Msi2isisTestCubicConvolution
 3/13 Test #1936: Msi2isis.Msi2isisTestCubicConvolution ..........   Passed    0.44 sec
      Start 1937: Msi2isis.Msi2isisTestNearestNeighbor
 4/13 Test #1937: Msi2isis.Msi2isisTestNearestNeighbor ...........   Passed    0.43 sec
      Start 1938: Msi2isis.Msi2isisTestBilinear
 5/13 Test #1938: Msi2isis.Msi2isisTestBilinear ..................   Passed    0.42 sec
      Start 1939: Msi2isis.Msi2isisTestAMissingLabelFile
 6/13 Test #1939: Msi2isis.Msi2isisTestAMissingLabelFile .........   Passed    0.14 sec
      Start 1940: Msi2isis.Msi2isisTestBInvalidInstrumentId
 7/13 Test #1940: Msi2isis.Msi2isisTestBInvalidInstrumentId ......   Passed    0.15 sec
      Start 1941: Msi2isis.Msi2isisTestCInvalidSize
 8/13 Test #1941: Msi2isis.Msi2isisTestCInvalidSize ..............   Passed    0.16 sec
      Start 1942: Msi2isis.Msi2isisTestDInvalidSize2
 9/13 Test #1942: Msi2isis.Msi2isisTestDInvalidSize2 .............   Passed    0.15 sec
      Start 1943: Msi2isis.Msi2isisTestEInvalidSampleDirection
10/13 Test #1943: Msi2isis.Msi2isisTestEInvalidSampleDirection ...   Passed    0.15 sec
      Start 1944: Msi2isis.Msi2isisTestFInvalidLineDirection
11/13 Test #1944: Msi2isis.Msi2isisTestFInvalidLineDirection .....   Passed    0.16 sec
      Start 1945: Msi2isis.Msi2isisTestGInvalidProjectedImage
12/13 Test #1945: Msi2isis.Msi2isisTestGInvalidProjectedImage ....   Passed    0.16 sec
      Start 1946: Msi2isis.Msi2isisTestHLabelTranslateError
13/13 Test #1946: Msi2isis.Msi2isisTestHLabelTranslateError ......   Passed    0.33 sec

100% tests passed, 0 tests failed out of 13

Total Test time (real) =   3.51 sec
(Mar282022) zion2[1375]: ctest -R NearMsiCameraTest
Test project /Users/kbecker/ISIS/Isis4Dev/Eros/MSI/Mar282022/ISIS3/build
    Start 2079: NearMsiCameraCube.NearMsiCameraTest
1/1 Test #2079: NearMsiCameraCube.NearMsiCameraTest ...   Passed    0.20 sec

100% tests passed, 0 tests failed out of 1

Total Test time (real) =   0.36 sec
```
## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation change (update to the documentation; no code change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] I have read and agree to abide by the [Code of Conduct](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/Code-Of-Conduct.md)
- [x] I have read the [**CONTRIBUTING**](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CONTRIBUTING.md) document.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] I have added myself to the [.zenodo.json](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/.zenodo.json) document.
- [x] I have added any user impacting changes to the [CHANGELOG.md](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CHANGELOG.md) document.

## Licensing
This project is mostly composed of free and unencumbered software released into the public domain, and we are unlikely to accept contributions that are not also released into the public domain. Somewhere near the top of each file should have these words:

> This work is free and unencumbered software released into the public domain. In jurisdictions that recognize copyright laws, the author or authors of this software dedicate any and all copyright interest in the software to the public domain.

- [x] I dedicate any and all copyright interest in this software to the public domain. I make this dedication for the benefit of the public at large and to the detriment of my heirs and successors. I intend this dedication to be an overt act of relinquishment in perpetuity of all present and future rights to this software under copyright law.

KrisBecker committed:
Rename NEAR/MSI main.cpp to msi2isis.cpp

Rename the NEAR/MSI ingestion app main.cpp to msi2isis.cpp to covert to callable function

KrisBecker committed:
Remove old msi2isis tests in tsts directory

The old tests have been replaced by new FunctionalTests

KrisBecker committed:
Update msi2isis to support new datasets

msi2isis has been modified to accept expanded datasets

* Accept images that have already been expanded to correct aspect (square pixels)
* New TRIM option added to optional turn off trimming of borders
* Modify SCLK start and stop times to be compatible with NAIF SCLK kernel format. This provides compatiblity for the sumspice application
* Save original start/stop SCLK values to new keywords OriginalSpacecraftClockStartCount/
OriginalSpacecraftClockStopCount
* Convert the app to a callable function

KrisBecker committed:
Add msi2isis functional tests and data

msi2isis funtional test and data

* Converted all old app tests to new functional tests
* Added label and data files
* msi2isis has accompanying .lbl files as input but .gitignore excludes these files. Add explicit ignores for these *.lbl files

KrisBecker committed:
Update MsiCamera model to support sumspice

MsiCamera model modified so that sumspice works correcty

* Use SpacecraftClockStartCount rather than StartTime to determine observation times for sumspice compatibility
* Add IAKs to track in repository
* Add light time and observer parameters to IAK version 2
* Bump camera version in plugin file due to new keywords loaded at spiceinit time

KrisBecker committed:
Add MsiCamera gtest to replace old test

MsiCamera Gtest written to replace old version

* Add new MsiCamera tests and Fixtures

KrisBecker committed:
Add optional keyword in serial number generation

NearMsiSerialNumber.trn update to add option keyword ControlSN to support use of simulated image in control processes

KrisBecker committed:
Removed MsiCamera unit test

msiCamera unitTest.cpp in favor of NearMsiCameraCube.NearMsiCameraTest

jessemapel commented:
Gonna get started on reviewing this, but I'll respond to a few things up front:

>  Of note is I had to explicitly add (unignore) the NEAR/MIS *.lbl files to .gitignore as lbl files are ignored repo-wide.

This is intentional, we don't want binary files accidentally added to the git repo as it makes for extremely large diffs and bloats the repo.

> NOTE that I did not update the application test as I am unsure how to get these changes into the system. I think it only requires an update to the truth files once the tests have been rerun. Please advise/assist.

Which application tests are you talking about? The old msi2isis app tests should be removed as part of this PR.

> I have added any user impacting changes to the [CHANGELOG.md](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CHANGELOG.md) document.

Be sure to add a changelog entry under UNRELEASED / ADDED

KrisBecker commented:
> Which application tests are you talking about? The old msi2isis app tests should be removed as part of this PR.

Actually this is the mission module tests in $ISISROOT/src/near.

```
(Mar282022) zion2[1377]: ctest -R near_module_test_msi
Test project /Users/kbecker/ISIS/Isis4Dev/Eros/MSI/Mar282022/ISIS3/build
    Start 1103: near_module_test_msi
1/1 Test #1103: near_module_test_msi .............***Failed   15.74 sec

0% tests passed, 1 tests failed out of 1

Label Time Summary:
app    =  15.74 sec*proc (1 test)

Total Test time (real) =  15.90 sec

The following tests FAILED:
	1103 - near_module_test_msi (Failed)
Errors while running CTest
Output from these tests are in: /Users/kbecker/ISIS/Isis4Dev/Eros/MSI/Mar282022/ISIS3/build/Testing/Temporary/LastTest.log
Use "--rerun-failed --output-on-failure" to re-run the failed cases verbosely.
```
I looked at the output and it appears these test require updating test data and I am uncertain how to make that change from my end.

KrisBecker committed:
Add updates to CHANGELOG.md

jessemapel commented:
I can get those tests updated once this PR is merged.

jessemapel commented on CHANGELOG.md:
```diff
@@ -38,6 +38,7 @@ release.
 ### Changed
 
 ### Added
+- Improved functionality of msi2isis and MsiCamera model to support new Eros dataset, including support for Gaskell's SUMSPICE files that adjust timing, pointing and spacecraft position ephemeris. [#4886] (https://github.com/USGS-Astrogeology/ISIS3/issues/4886)
```
Remove the space between the `[]` and `()` in the issue link

KrisBecker replied:
Done

KrisBecker commented:
Just an FYI, after all the ctests have been ran on my system, there are some residual files that consistently remain (or marked as untracked).  These files are:
```
isis/tests/data/isisimport/pds4.xml
isis/tests/winnowGuilty.csv
isis/tests/winnowIgnored.csv
```

jessemapel commented on isis/src/near/apps/msi2isis/msi2isis.cpp:
```diff
@@ -0,0 +1,323 @@
+    else {
+        // Image has the proper expanded size
+        enlargeProcessOutCube = importProcessOutCube;
+        // inatt.addAttributes(enlargeProcessOutCube);
```
What's this commented line for?

KrisBecker replied:
Removed.

jessemapel commented on isis/src/near/apps/msi2isis/msi2isis.cpp:
```diff
@@ -0,0 +1,323 @@
+    else {  // TRIM == FALSE
+        // Effectively turns off trimming
+      trim = 0;          // No grim lines/samples
+    }
```
Just do
```
if( !ui.GetBoolean("TRIM") ) {
  trim = 0;
}
```
No need to include the default case in the if/else.

KrisBecker replied:
Done.

jessemapel commented on isis/src/near/apps/msi2isis/msi2isis.cpp:
```diff
@@ -0,0 +1,323 @@
+    }
+    else {
+        // Image has the proper expanded size
+        enlargeProcessOutCube = importProcessOutCube;
```
Do we want to also disable trimming here? It seems like you either:

* have an old corrected image and want to expand + trim

or

* have a new corrected image want to leave it alone

KrisBecker replied:
I would rather let the user decide use of this option.

We have found it useful in to apply trimming to new data (or vice versa) for direct comparisons in maps to evaluate the effects trimmed/untrimmed edges. Without this option, users would have to run trim for exactly 33 pixels on all sides, which clusters up workspace and complicates scripts.

jessemapel commented on isis/tests/Fixtures.cpp:
```diff
@@ -1693,4 +1693,14 @@ namespace Isis {
+
+    QString msicube = "m0155881376f3_2p_cif_dbl.cub";
+    QString testPath = tempDir.path() + "/" + msicube;
+    QFile::copy("data/near/msicamera/"+msicube, testPath);
```
Why does the input data need to be copied out of the data directory?

KrisBecker replied:
This has been cleaned up with improved implementation in the Fixture.

jessemapel commented on isis/tests/FunctionalTestsMsi2isis.cpp:
```diff
@@ -0,0 +1,469 @@
+
+TEST(Msi2isis, Msi2isisTestGblIngestDefault) {
+  Pvl appLog;
+  QTemporaryDir prefix;
```
Use the `TempTestingFiles` fixture instead of making your own temp directory. This makes it easier for us to modify how we handle test outputs and temp files in the future.

KrisBecker replied:
Done. This changes the test name patterns a bit:
```
(Mar282022) zion2[1589]: ctest -R Msi2isis
Test project /Users/kbecker/ISIS/Isis4Dev/Eros/MSI/Mar282022/ISIS3/build
      Start 1710: TempTestingFiles.Msi2isisTestGblIngestDefault
 1/13 Test #1710: TempTestingFiles.Msi2isisTestGblIngestDefault ..........   Passed    0.56 sec
      Start 1711: TempTestingFiles.Msi2isisTestGblIngestNoTrim
 2/13 Test #1711: TempTestingFiles.Msi2isisTestGblIngestNoTrim ...........   Passed    0.34 sec
      Start 1712: TempTestingFiles.Msi2isisTestCubicConvolution
 3/13 Test #1712: TempTestingFiles.Msi2isisTestCubicConvolution ..........   Passed    0.50 sec
      Start 1713: TempTestingFiles.Msi2isisTestNearestNeighbor
 4/13 Test #1713: TempTestingFiles.Msi2isisTestNearestNeighbor ...........   Passed    0.47 sec
      Start 1714: TempTestingFiles.Msi2isisTestBilinear
 5/13 Test #1714: TempTestingFiles.Msi2isisTestBilinear ..................   Passed    0.49 sec
      Start 1715: TempTestingFiles.Msi2isisTestAMissingLabelFile
 6/13 Test #1715: TempTestingFiles.Msi2isisTestAMissingLabelFile .........   Passed    0.18 sec
      Start 1716: TempTestingFiles.Msi2isisTestBInvalidInstrumentId
 7/13 Test #1716: TempTestingFiles.Msi2isisTestBInvalidInstrumentId ......   Passed    0.18 sec
      Start 1717: TempTestingFiles.Msi2isisTestCInvalidSize
 8/13 Test #1717: TempTestingFiles.Msi2isisTestCInvalidSize ..............   Passed    0.17 sec
      Start 1718: TempTestingFiles.Msi2isisTestDInvalidSize2
 9/13 Test #1718: TempTestingFiles.Msi2isisTestDInvalidSize2 .............   Passed    0.18 sec
      Start 1719: TempTestingFiles.Msi2isisTestEInvalidSampleDirection
10/13 Test #1719: TempTestingFiles.Msi2isisTestEInvalidSampleDirection ...   Passed    0.17 sec
      Start 1720: TempTestingFiles.Msi2isisTestFInvalidLineDirection
11/13 Test #1720: TempTestingFiles.Msi2isisTestFInvalidLineDirection .....   Passed    0.17 sec
      Start 1721: TempTestingFiles.Msi2isisTestGInvalidProjectedImage
12/13 Test #1721: TempTestingFiles.Msi2isisTestGInvalidProjectedImage ....   Passed    0.17 sec
      Start 1722: TempTestingFiles.Msi2isisTestHLabelTranslateError
13/13 Test #1722: TempTestingFiles.Msi2isisTestHLabelTranslateError ......   Passed    0.36 sec

100% tests passed, 0 tests failed out of 13

Total Test time (real) =   4.10 sec
```

jessemapel commented on isis/tests/FunctionalTestsMsi2isis.cpp:
```diff
@@ -0,0 +1,469 @@
+
+  }
+  catch (IException &e) {
+    FAIL() << "Unable to ingest NEAR/MSI GBL file: " << e.toString().toStdString().c_str() << std::endl;
```
No need to add `c_str`, gtest can take standard strings.

KrisBecker replied:
Done.

jessemapel commented on isis/tests/FunctionalTestsMsi2isis.cpp:
```diff
@@ -0,0 +1,469 @@
+
+ // Kernels Group
+  PvlGroup &kernel = isisLabel->findGroup("Kernels", Pvl::Traverse);
+  ASSERT_EQ(int(kernel["NaifFrameCode"]), -93001);
```
There's no reason to re-test all of these label values. Only check for things that change due to `trim=false`.

jessemapel commented on isis/tests/FunctionalTestsMsi2isis.cpp:
```diff
@@ -0,0 +1,469 @@
+  ASSERT_NEAR(hist->Sum(), 8997.9670307614, .0001);
+  ASSERT_EQ(hist->ValidPixels(), 221244);
+  ASSERT_EQ(hist->NullPixels(), 0);
+  ASSERT_NEAR(hist->StandardDeviation(), 0.017675184374796, .0001);
```
Instead of testing against hard coded values, actually check that the outside pixels are properly set to NULL and the inside pixels are the same as before.

Something we do often with tests like this is to run once with the parameter set to true and again with the parameter set to false. Then, just check that the differences between the two are what we expect. So, here you can use a portal to go through the trimmed and un-trimmed cubes and check that the interiors are the same and the outer edge of the trimmed cube is all set to NULL.

KrisBecker replied:
Done.

jessemapel commented on isis/tests/FunctionalTestsMsi2isis.cpp:
```diff
@@ -0,0 +1,469 @@
+  ASSERT_NEAR(hist->Sum(), 6947.8146519054, .0001);
+  ASSERT_EQ(hist->ValidPixels(), 162966);
+  ASSERT_EQ(hist->NullPixels(), 58278);
+  ASSERT_NEAR(hist->StandardDeviation(), 0.016649588425538, .0001);
```
Most of these ASSERTs can be changed to EXPECTs. Generally, I only use a hard ASSERT when it would result in the following test code to fail. The most common examples of this are testing the size of output before iterating over it and checking that a PVL group/object exists before checking its contents.

KrisBecker replied:
Done.

jessemapel commented on isis/tests/FunctionalTestsMsi2isis.cpp:
```diff
@@ -0,0 +1,469 @@
+
+ // Kernels Group
+  PvlGroup &kernel = isisLabel->findGroup("Kernels", Pvl::Traverse);
+  ASSERT_EQ(int(kernel["NaifFrameCode"]), -93001);
```
Again you don't need to re-test all of these label values, just anything that changes when you set cubic convolution.

KrisBecker replied:
I have left these tests in because if for some reason the input changes, the test will certainly fail. Testing the label will immediately indicate why.

jessemapel replied:
The problem is that these lines are re-checking that same values as previous tests but not doing anything that should change them. For each input, we should only check that their labels are correct one time. So, it is okay to have one test check that the old label is correct and one test check that the new label is correct. We should not be checking that in every test that uses them.

KrisBecker replied:
Done.

jessemapel commented on isis/tests/FunctionalTestsMsi2isis.cpp:
```diff
@@ -0,0 +1,469 @@
+   UserInterface options(dAPP_XML, args);
+
+   // TEST A: pass in a fits file without label
+   ASSERT_ANY_THROW(msi2isis(options, &appLog));
```
We use a more complex method to test exceptions that checks the message and error type is what we expect.

see https://github.com/USGS-Astrogeology/ISIS3/wiki/Writing-ISIS3-Tests-Using-Gtest-and-Ctest#testing-exceptions

You can also use an `EXPECT_THAT` to check the substring like this https://github.com/USGS-Astrogeology/ISIS3/blob/dev/isis/tests/FunctionalTestsApolloremrx.cpp#L119

KrisBecker replied:
Done.

jessemapel commented on isis/tests/Fixtures.cpp:
```diff
@@ -1693,4 +1693,14 @@ namespace Isis {
     }
   }
+
+void NearMsiCameraCube::Init() {
```
Instead of adding a new Init method, override the parent SetUp method and GTest will automatically call this.

KrisBecker replied:
Fixed.

jessemapel commented on isis/tests/Fixtures.cpp:
```diff
@@ -1693,4 +1693,14 @@ namespace Isis {
+    QString msicube = "m0155881376f3_2p_cif_dbl.cub";
+    QString testPath = tempDir.path() + "/" + msicube;
+    QFile::copy("data/near/msicamera/"+msicube, testPath);
+    testCube = new Cube(testPath, "rw");
```
You need to add a `TearDown` method to clean this up

KrisBecker replied:
Fixed.

jessemapel commented on isis/tests/NearMsiCameraTests.cpp:
```diff
@@ -0,0 +1,67 @@
+using namespace Isis;
+
+TEST_F(NearMsiCameraCube, NearMsiCameraTest) {
+  Init();
```
See my comments on the test fixture, GTest will call this for you if you make it SetUp.

KrisBecker replied:
Fixed.

jessemapel commented on isis/tests/Fixtures.cpp:
```diff
@@ -1693,4 +1693,14 @@ namespace Isis {
+void NearMsiCameraCube::Init() {
+    TempTestingFiles::SetUp();
+
+    QString msicube = "m0155881376f3_2p_cif_dbl.cub";
```
This doesn't appear to be in the PR. We also usually don't store full cubes for tests. We generate an ISD and store that + the cube label. Then, we generate the cube at run time.

KrisBecker replied:
How do I generate an ISD for a NEAR/MSI image?

For now, I have unignored this cube in the repo.

jessemapel replied:
It may require some extra steps. I'll spend some time tomorrow seeing if I can do it with the cube you pushed. Can you make sure you have the "allow edits and access to secrets from maintainers" box checked and I can push it up.
![Screen Shot 2022-04-05 at 4 04 12 PM](https://user-images.githubusercontent.com/22942817/161865119-afb7f404-47da-4fec-a470-e9a0dcebf358.png)

KrisBecker replied:
It is checked.

Thanks...

jessemapel reviewed (commented):
The code changes look good. I added a bunch of comments about how to improve the tests.

KrisBecker committed:
Add NEAR/MSI spiced cube for MsiCamera unit test

The MsiCamera requires a spiceinit cube, but .gitignore excludes .cub files. This commit addresses this issue by explicitly unignoring the Eros cube used for the test.

KrisBecker committed:
Fix issues from review comments

* Removed space between issue number and url in CHANGELOG.md

msi2isis
* Remove commented code
* Simplify trimming. Let user decide uses for this option rather than assume it should not be used under certain conditions

KrisBecker committed:
Cleaned up NearMsiCameraCube Fixture

jessemapel commented on isis/tests/Fixtures.cpp:
```diff
@@ -1694,13 +1694,13 @@ namespace Isis {
-    QFile::copy("data/near/msicamera/"+msicube, testPath);
-    testCube = new Cube(testPath, "rw");
-
+  void NearMsiCameraCube::TearDown() {
```
Now that this is a smart pointer, you actually don't need this anymore

KrisBecker replied:
I was uncertain about that because the Cube object will be deleted at the end of the test. What happens in TearDown()? Does the temp directory/files get removed at that point? If so, letting the pointer go out of scope when the class is destroyed seems too late to delete the Cube pointer, which I see most TearDown() methods doing.

jessemapel replied:
Yeah the parent TearDown for the test fixture will get called which removes the temp directory. It should be okay to just allow the smart pointer to clean up at that point; I don't think the cube file needs to still exist when the cube is destroyed.

KrisBecker replied:
Seems ill-advised as it might limit what you can change in future mods, but OK. I will leave this in.

I think I should also explicitly call the inherited class' TearDown() otherwise it won't do its thing since it is overridden?

jessemapel replied:
We don't need to call the parent teardown because it doesn't have one, we just let the normal member clean-up remove the tempdir.

KrisBecker replied:
Ok, I see the difference from other fixtures. Thanks.

jessemapel committed:
Moved NEAR MSI test to use label and ISD

jessemapel commented:
I've moved the test cube to a label + ISD setup, it's not a huge savings here because the original cube is ~1MB and the label + ISD is ~15KB, but for larger images this is a big deal.

I'm going to get the SPICE updated in our system now.

KrisBecker committed:
Address suggested improvements to msi2isis tests

KrisBecker committed:
Merge branch 'eros_msi_pdart_mods' of https://github.com/KrisBecker/ISIS3 into eros_msi_pdart_mods

KrisBecker commented:
Just pushed test modifications that address your comments/concerns.

Please let me know if there is anything remaining preventing a merge.

Thanks...

jessemapel commented on isis/tests/FunctionalTestsMsi2isis.cpp:
```diff
@@ -2,104 +2,168 @@
-  ASSERT_EQ(inst["SpacecraftClockStartCount"][0].toStdString(), "157063318856" );
-  ASSERT_EQ(inst["SpacecraftClockStopCount"][0].toStdString(), "157063318919" );
-  ASSERT_EQ(inst["OriginalSpacecraftClockStartCount"][0].toStdString(), "157063318.856" );
-  ASSERT_EQ(inst["OriginalSpacecraftClockStopCount"][0].toStdString(), "157063318.919" );
```
You want to keep the conversion to a c++ string, you just don't need to go to a c string. googletest is able to compare QStrings, but it is not able to report differences in an intelligible way. You end up with some hex byte output that is hard to interpret.

KrisBecker replied:
Added back conversions toStdString().

jessemapel commented on isis/tests/FunctionalTestsMsi2isis.cpp:
```diff
@@ -2,104 +2,168 @@
+  EXPECT_EQ(innerImage.ValidPixels(), n_valid);
+
+  EXPECT_EQ(trimEdge.Sum(), 0.0);
+  EXPECT_EQ(trimEdge.NullPixels(), hist->NullPixels());
```
Instead of comparing the number of null pixels in the edge to the number of null pixels in the entire image, compare the total number of pixels in the edge to the number of null pixels in the edge, they should be equal.

jessemapel commented on isis/tests/FunctionalTestsMsi2isis.cpp:
```diff
@@ -2,104 +2,168 @@
+        }
+        else {
+          sum += lbuf[s];
+          n_valid++;
```
These counts are redundant and just duplicating/testing histogram's functionality. Let the ImageHistogram class handle this for you.

jessemapel commented on isis/tests/FunctionalTestsMsi2isis.cpp:
```diff
@@ -2,104 +2,168 @@
+  EXPECT_EQ(innerImage.Average(), hist->Average());
+  EXPECT_EQ(innerImage.Sum(), hist->Sum());
+  EXPECT_EQ(innerImage.ValidPixels(), hist->ValidPixels());
+  EXPECT_EQ(innerImage.StandardDeviation(), hist->StandardDeviation());
```
If we are somehow overtrimming the image, for example we trim 34 instead of 33 pixels, then these expectations will still all be satisfied. Instead compare the inside of the trimmed image to the inside of an untrimmed image.

KrisBecker replied:
Done.

jessemapel commented on isis/tests/FunctionalTestsMsi2isis.cpp:
```diff
@@ -111,61 +175,128 @@ TEST(Msi2isis, Msi2isisTestGblIngestNoTrim) {
+  EXPECT_EQ(composite_image_null_pixels, hist->NullPixels());
+
+  EXPECT_EQ(trimEdge.NullPixels(), 0);
+  EXPECT_EQ(trimEdge.NullPixels(), n_nulls);
```
If you compare the trimmed and untrimmed images in the previous test, this whole test can go away.

KrisBecker replied:
Done.

Let me know if this update resolves your concerns.

KrisBecker committed:
Updated tests after feedback from PR review

jessemapel committed:
Merge branch 'dev' into eros_msi_pdart_mods
